hw/mcu/microchip/pic32mz/src/hal_uart.c

```diff
@@ -0,0 +1,736 @@
+            ret += pps_configure_input(uarts[port].u_pins->rx, U3RX_IN_FUNC);
+            if (flow_ctl == HAL_UART_FLOW_CTL_RTS_CTS) {
+                ret += pps_configure_output(uarts[port].u_pins->rts, U3RTS_OUT_FUNC);
+                if (uarts[port].u_pins->cts != PIN_UNUSED) {
+                    ret += pps_configure_input(uarts[port].u_pins->cts, U3CTS_IN_FUNC);
+                }
+            }
+            break;
+#endif
+#if MYNEWT_VAL(UART_3)
+        case 3:
+            ret += pps_configure_output(uarts[port].u_pins->tx, U4TX_OUT_FUNC);
+            ret += pps_configure_input(uarts[port].u_pins->rx, U4RX_IN_FUNC);
+            if (flow_ctl == HAL_UART_FLOW_CTL_RTS_CTS) {
+                ret += pps_configure_output(uarts[port].u_pins->rts, U4RTS_OUT_FUNC);
+                if (uarts[port].u_pins->cts != PIN_UNUSED) {
+                    ret += pps_configure_input(uarts[port].u_pins->cts, U4CTS_IN_FUNC);
+                }
+            }
+            break;
+#endif
+#if MYNEWT_VAL(UART_4)
+        case 4:
+            ret += pps_configure_output(uarts[port].u_pins->tx, U5TX_OUT_FUNC);
+            ret += pps_configure_input(uarts[port].u_pins->rx, U5RX_IN_FUNC);
+            if (flow_ctl == HAL_UART_FLOW_CTL_RTS_CTS) {
+                ret += pps_configure_output(uarts[port].u_pins->rts, U5RTS_OUT_FUNC);
+                if (uarts[port].u_pins->cts != PIN_UNUSED) {
+                    ret += pps_configure_input(uarts[port].u_pins->cts, U5CTS_IN_FUNC);
+                }
+            }
+            break;
+#endif
+#if MYNEWT_VAL(UART_5)
+        case 5:
+            ret += pps_configure_output(uarts[port].u_pins->tx, U6TX_OUT_FUNC);
+            ret += pps_configure_input(uarts[port].u_pins->rx, U6RX_IN_FUNC);
+            if (flow_ctl == HAL_UART_FLOW_CTL_RTS_CTS) {
+                ret += pps_configure_output(uarts[port].u_pins->rts, U6RTS_OUT_FUNC);
+                if (uarts[port].u_pins->cts != PIN_UNUSED) {
+                    ret += pps_configure_input(uarts[port].u_pins->cts, U6CTS_IN_FUNC);
+                }
+            }
+            break;
+#endif
+        default:
+            return -1;
+        }
+        if (ret) {
+            return -1;
+        }
+    }
+
+    /* Set pin as digital input to clear ANSEL bit if any */
+    hal_gpio_init_in(uarts[port].u_pins->rx, HAL_GPIO_PULL_NONE);
+
+    divisor = peripheral_clk / (4 * baudrate) - 1;
+
+    /* disable */
+    UxMODE(port) = 0;
+    _nop();
+    UxBRG(port) = divisor;
+    UxMODE(port) = mode;
+    UxSTA(port) = _U1STA_URXEN_MASK | _U1STA_UTXEN_MASK;
+
+    switch (port) {
+#if MYNEWT_VAL(UART_0)
+    case 0:
+        /* clear RX interrupt flag */
+        IFS3CLR = _IFS3_U1RXIF_MASK;
+
+        /* enable RX interrupt */
+        IEC3SET = _IEC3_U1RXIE_MASK;
+
+        /* set rx interrupt priority */
+        IPC28CLR = _IPC28_U1RXIP_MASK;
+        IPC28SET = (1 << _IPC28_U1RXIP_POSITION); // priority 1
```

Comment:
The section below uses line comments which ain't really part of our coding style (I am totally fine with it myself)...

Reply:
done